### ISSUE TITLE
Disk space not released unless NO transaction exists

### ISSUE THREAD

**Issue by swingbit** [MonetDB/MonetDB#7349]
**Describe the bug**
Disk space of dropped tables is only released when no transaction exists, no matter when the transactions were started or what they do.

**To Reproduce**
With an empty DB, open 3 mclient sessions, and somewhere else keep a `watch du -sh <dbfolder>`.
Reproduce the following sequence:

| sequence | Session 1 | Session 2 | Session 3 | `du -sh <dbfolder>` (after command) |
| ---- | --------- | --------- | --------- | --------- |
| 0    | | | | 4K |
| 1    | `create table t as select * from generate_series(1, 10 * 1000 * 1000);` | | | 45M |
| 2    | | `start transaction`; | | 45M |
| 3    | `drop table t`; | | | 45M |
| 4    | | | `start transaction;` | 45M |
| 5    | | `rollback;` | | 45M |
| 6    | | | `rollback;` | 4K |

The `rollback` at time 5 should allow disk space to be released, because the only transaction that starts before the `DROP TABLE t` ends. I waited for very long after issuing the `rollback`, but nothing happened. 

Instead, disk space is released only when NO transactions exist, i.e. at time 6 (after few seconds from the `rollback`). 

Note: If you keep opening a new transaction before the last active one ends, thus never leaving a moment with no transactions active, then disk space from table `t` is never released.

**Expected behavior**
Disk space is released after the `rollback` at time 5.

**Software versions**
 - MonetDB version number 11.43.14
 - OS and version: Fedora 37
 - self-installed and compiled

**Comment by aris-koning** [comment]
Thanks I am working on an improved version of the GC on the transactional level. I will take this into account as a nice test case.

**Comment by njnes** [comment]
Your table misses some more details. 
Once your table gets dropped and all other transactions (older than the drop) finished, the bat(s) of this table could still be in the gdk_logger's catalog_bid bat. Only once the 'wal' file containing the drop is 'flushed/rotated', the bat is finally transient and with the next (gdk level) commit removed from the disc.

**Comment by njnes** [comment]
The wal file rotation can only be done if a new one has been started. Starting too many isn't efficient, ie we have a fixed size of 2G for these files. If the server is inactive for some time the new file is created independent of previous wal file 
size.

**Comment by swingbit** [comment]
@njnes Thanks for the additional details.

In the example above, I've waited up to 30 minutes, but still nothing happened.

**Comment by swingbit** [comment]
Is Jun2023 expected to change this behaviour?

I'm replicating now the same experiment on Jun2023, and at first sight it seems to behave worse.

The original expectation was a disk space release at time 5. Jan2022 (at the time of the bug report) was releasing at time 6. What I observe now is that the disk space is not released at all, *not even after closing all the existing connections* and waiting for several minutes with the server completely idle. The only way is to restart the server.

I even tried with 1000x more data, for a total allocation of 76G, hoping that this would trigger the GC more eagerly. No difference.

UPDATE: the disk space is not released even after 1 hour with no connections and no activity.

The status of the `sql_logs` folder:
```
$ ll dbfarms/dbfarm.Spinque_Jun2023.preview/x/sql_logs/sql/
total 8
-rw-------. 1 roberto roberto 266 May 30 15:56 log
-rw-------. 1 roberto roberto  88 May 30 15:59 log.3
-rw-------. 1 roberto roberto   0 May 30 16:03 log.4
```

Can you please try to replicate this on your side? It seems a pretty serious issue to me.

**Comment by swingbit** [comment]
I'll summarise the recent developments to hopefully close the loop. @njnes please correct me if I'm wrong.

The issue found in the last comment, i.e. that disk space would never be released, not even closing all sessions, was fixed by commit d522d3570e3bd6399ea332d91f56d95031a14389.

The original issue, i.e. that disk space would not be released until no active transaction exists, has been fixed by commit bc1554d6f91a5b02eb9921a995ae642cb36d24b6.

My test above still fails as is, but it succeeds if one more session is used in the chain of transactions. I understood that this is expected due internal reasons, which is fine by me.
